terminal: Introduce themes

marusak opened the pull request:
This is the first version of themes in terminal.
There are a few questions which needs to be answered, such as:

1.  What is default theme?
2. How should the selection look like? *
3. What themes we want and are these suitable?

* This is discussed in #11177 @garrett has some awesome ideas for simple dialog-thing  which would contain nice theme selection as well as setting of the font size. Question is if this should be in this PR or as follow-up.

Any comments are welcome. Now default theme is dark.
Screenshots:
![darkmodedone](https://user-images.githubusercontent.com/12330670/52840084-7be2e200-30f8-11e9-8bbc-0869de1543ca.png)
![lightmodedone](https://user-images.githubusercontent.com/12330670/52840085-7c7b7880-30f8-11e9-8ea3-cab527703dfa.png)

Fixes #11177

Marking as no-test for now, if this will be close to landing I will enable tests (I dont want to test it pointlessly if big changes will be required)

marusak commented:
Rebased and looks nice with context menu :) 
![screenshot from 2019-02-15 08-45-23](https://user-images.githubusercontent.com/12330670/52842026-1560c280-30fe-11e9-978d-7b9c1b6d7b0a.png)
![screenshot from 2019-02-15 08-45-09](https://user-images.githubusercontent.com/12330670/52842027-15f95900-30fe-11e9-9320-8fa0a283aa79.png)

martinpitt commented:
The two current themes from that are from solarized? They look reasonably tasteful indeed, but they seem to lose most colors. E. g. my `ls` output is now just black/white and two shades of grey. I'd still like to retain high-contrast/high-color theme, and our current one as a default (i. e. proper black background instead of the blue-greyish one).

How about Black (default), Dark, Light, White?

martinpitt commented on pkg/systemd/terminal.css:
```diff
@@ -13,9 +13,27 @@ html, body {
+
+.light {
+    background-color: #fdf6e3;
+}
```
I take it we can't put these into the themes because of CSP? Can these be prefixed with `theme-`, to avoid overlaps with other classes when our theme list grows?

martinpitt commented on pkg/systemd/terminal.jsx:
```diff
@@ -51,6 +54,11 @@ const _ = cockpit.gettext;
 
+        onThemeChanged(value) {
+            this.setState({ theme: value });
+            cockpit.localStorage.setItem('terminal-theme', value);
```
This doesn't seem  to work for me in Firefox. I switched to light, logged out, logged back in, and ended up with dark again.

martinpitt commented on pkg/systemd/terminal.jsx:
```diff
@@ -71,19 +79,28 @@ const _ = cockpit.gettext;
-                        <button ref="resetButton"
-                             className="btn btn-default pull-right"
-                             onClick={this.onResetClick}>{_("Reset")}</button>
+                        <div>
```
Why does this need an extra div wrapper?

marusak replied:
Because of flex. I want terminal title to be on left and these two buttons to be on right and flex can nicely spread them. However I want to keep buttons together. Probably this can be done in more elegant way, but someone with more skills with css would need to give me some hint :)

martinpitt replied:
Ack, thanks! Fine then.

martinpitt commented on pkg/systemd/terminal.jsx:
```diff
@@ -71,19 +79,28 @@ const _ = cockpit.gettext;
+                        </div>
                     </div>
-                    <div className="panel-body">
+                    <div className={["panel-body", this.state.theme].join(' ')}>
```
Please leave an extra space after { and before }. Also, `"panel-body " + this.state.theme` is slightly simpler.

martinpitt commented on test/verify/check-terminal:
```diff
@@ -161,6 +161,25 @@ PROMPT_COMMAND='printf "\\033]0;%s@%s:%s\\007" "${USER}" "${HOSTNAME%%.*}" "${PW
+        b.wait_attr(".xterm-viewport", "style", light_style)
+
+        # Reload and white-style should be remembered
+        # Reload on different page, as reloading on terminal prompts asking if you want to leave the site
```
You can also just log out and back in, which is a more realistic and demanding test.

marusak replied:
Ok, I changed it to relogin, but still needs to do it on different page, as on terminal page I am asked if I want to leave the page. (interesting fact - This happens only in test, normally on logout I am not asked is I want to leave the page). It is most likely possible to do even on the same page, but I think it is good enough to switch page.

marusak commented:
> They look reasonably tasteful indeed, but they seem to lose most colors. E. g. my ls output is now just black/white and two shades of grey.

This is unwanted behavior. See my screenshots, I execute there `ls` and it has normal colors. I believe those two shades of grey are "light" version of colors, which are usually used for bold text. Is it possible you used `dircolors` command in past? Or what is your `LS_COLORS` env variable?

```
echo $LS_COLORS
rs=0:di=38;5;33:ln=38;5;51:mh=00:pi=40;38;5;11:so=38;5;13:do=38;5;5:bd=48;5;232;38;5;11:cd=48;5;232;38;5;3:or=48;5;232;38;5;9:mi=01;05;37;41:su=48;5;196;38;5;15:sg=48;5;11;38;5;16:ca=48;5;196;38;5;226:tw=48;5;10;38;5;16:ow=48;5;10;38;5;21:st=48;5;21;38;5;15:ex=38;5;40:*.tar=38;5;9:*.tgz=38;5;9:*.arc=38;5;9:*.arj=38;5;9:*.taz=38;5;9:*.lha=38;5;9:*.lz4=38;5;9:*.lzh=38;5;9:*.lzma=38;5;9:*.tlz=38;5;9:*.txz=38;5;9:*.tzo=38;5;9:*.t7z=38;5;9:*.zip=38;5;9:*.z=38;5;9:*.dz=38;5;9:*.gz=38;5;9:*.lrz=38;5;9:*.lz=38;5;9:*.lzo=38;5;9:*.xz=38;5;9:*.zst=38;5;9:*.tzst=38;5;9:*.bz2=38;5;9:*.bz=38;5;9:*.tbz=38;5;9:*.tbz2=38;5;9:*.tz=38;5;9:*.deb=38;5;9:*.rpm=38;5;9:*.jar=38;5;9:*.war=38;5;9:*.ear=38;5;9:*.sar=38;5;9:*.rar=38;5;9:*.alz=38;5;9:*.ace=38;5;9:*.zoo=38;5;9:*.cpio=38;5;9:*.7z=38;5;9:*.rz=38;5;9:*.cab=38;5;9:*.wim=38;5;9:*.swm=38;5;9:*.dwm=38;5;9:*.esd=38;5;9:*.jpg=38;5;13:*.jpeg=38;5;13:*.mjpg=38;5;13:*.mjpeg=38;5;13:*.gif=38;5;13:*.bmp=38;5;13:*.pbm=38;5;13:*.pgm=38;5;13:*.ppm=38;5;13:*.tga=38;5;13:*.xbm=38;5;13:*.xpm=38;5;13:*.tif=38;5;13:*.tiff=38;5;13:*.png=38;5;13:*.svg=38;5;13:*.svgz=38;5;13:*.mng=38;5;13:*.pcx=38;5;13:*.mov=38;5;13:*.mpg=38;5;13:*.mpeg=38;5;13:*.m2v=38;5;13:*.mkv=38;5;13:*.webm=38;5;13:*.ogm=38;5;13:*.mp4=38;5;13:*.m4v=38;5;13:*.mp4v=38;5;13:*.vob=38;5;13:*.qt=38;5;13:*.nuv=38;5;13:*.wmv=38;5;13:*.asf=38;5;13:*.rm=38;5;13:*.rmvb=38;5;13:*.flc=38;5;13:*.avi=38;5;13:*.fli=38;5;13:*.flv=38;5;13:*.gl=38;5;13:*.dl=38;5;13:*.xcf=38;5;13:*.xwd=38;5;13:*.yuv=38;5;13:*.cgm=38;5;13:*.emf=38;5;13:*.ogv=38;5;13:*.ogx=38;5;13:*.aac=38;5;45:*.au=38;5;45:*.flac=38;5;45:*.m4a=38;5;45:*.mid=38;5;45:*.midi=38;5;45:*.mka=38;5;45:*.mp3=38;5;45:*.mpc=38;5;45:*.ogg=38;5;45:*.ra=38;5;45:*.wav=38;5;45:*.oga=38;5;45:*.opus=38;5;45:*.spx=38;5;45:*.xspf=38;5;45:
```

marusak commented on pkg/lib/cockpit-components-terminal.jsx:
```diff
@@ -24,6 +24,47 @@ import { Terminal as Term } from "xterm";
+    green: "#859900"
+};
+
+const dark_theme = Object.assign({
```
`theme_core` cannot be first, because first argument is target, that is overwritten. So the last overwrite is used. I guess making deep copy would solve this problem.

martinpitt replied:
The usual way to do this is `Object.assign({}, theme_core, { overrides .. })`. But no need to  re-push again for this, we can change this later when we need it.

marusak replied:
I will update this as I will re-push anyway

martinpitt commented:
I don't have any `$LS_COLORS`, I just have `alias ls='ls --color=auto'`.

marusak commented:
Unsetting LS_COLORS breaks colors in the same way as you described. I have no idea how I got this set up, but it happens automatically for me.

marusak commented:
I've just pushed some other commits - most of theme can be merged in the end, but for now keeping them separated as it is easier to drop commit when someone will be against :) 

I've added black and white theme. And also selection now is through colored circles (plan is to have also font size setting in the same dropdown menu where now theme picking is).

![whitetheme](https://user-images.githubusercontent.com/12330670/52950753-5f58db00-3380-11e9-987c-7a1b4ea3d8c1.png)
![blacktheme](https://user-images.githubusercontent.com/12330670/52950755-5f58db00-3380-11e9-87a7-816d05133499.png)

And I forgot to comment, but I've updated all of your comments @martinpitt

martinpitt commented on pkg/systemd/terminal.jsx:
```diff
@@ -51,6 +57,13 @@ const _ = cockpit.gettext;
+        onThemeChanged(value) {
+            this.setState({ theme: value });
+            var cookie = "CockpitTerminalTheme=" + encodeURIComponent(value) +
+                         "; path=/; expires=Sun, 16 Jul 3567 06:23:41 GMT";
```
Why now a cookie instead of localstorage? Because it's easier to manage by the user? (Also, what is this magic time?)

marusak replied:
LocalStorage gets clean on logout. https://github.com/cockpit-project/cockpit/blob/master/src/base1/cockpit.js#L2399
(Just now I looked properly on function `setItem` and it is possible to set it in localStorage that way that is survives logout). But Cockpit saves language selection in cookie and all the code is picked from there.
But now I am thinking I should probably use localStorage and save it in a way that survives logout.

martinpitt replied:
Don't get me wrong, I'm fine with a cookie. It's also more obvious for a user when searching/cleaning cookies.

martinpitt commented on pkg/systemd/terminal.jsx:
```diff
@@ -51,6 +57,13 @@ const _ = cockpit.gettext;
+            this.setState({ theme: value });
+            var cookie = "CockpitTerminalTheme=" + encodeURIComponent(value) +
+                         "; path=/; expires=Sun, 16 Jul 3567 06:23:41 GMT";
+            document.cookie = cookie;
```
Shouldn't that be `theme_cookie` or so, to be more specific? Note that we already have a login session cookie.

marusak replied:
Language selection is saved in cookie `CockpitLang`

martinpitt replied:
Right, but when introspecting `document` or just for avoiding silly name conflicts, it'd still be better to give it a more specific name.

martinpitt replied:
Nevermind me, `document.cookie` is a standard API.

martinpitt commented on pkg/lib/cockpit-components-select.jsx:
```diff
@@ -118,11 +119,15 @@ export class StatelessSelect extends React.Component {
 
+        let iconClass = "caret";
+        if (this.props.iconClass)
+            iconClass = this.props.iconClass;
```
Please shorten that to `const iconClass = this.props.iconClass || "caret";`

Also, please adjust the commit message to "lib: Enable setting custom icon in Select component"

martinpitt commented on pkg/systemd/terminal.jsx:
```diff
@@ -4,10 +4,16 @@ import React from "react";
+const buttonToState = {
+    "dark-theme": "darkTheme",
+    "light-theme": "lightTheme",
+};
```
This feels like busywork. Why can't the state be the theme name directly?

martinpitt commented on pkg/systemd/terminal.jsx:
```diff
@@ -51,17 +57,22 @@ const _ = cockpit.gettext;
             cockpit.user().done(function (user) {
                 this.setState({ user: user, channel: this.createChannel(user) });
             }.bind(this));
+            this.setState({ [buttonToState[this.state.theme]]: " active" });
```
TBH I liked the previous `{ theme: name }` better. There can only ever be one active theme, and this structure allows (in principle) multiple themes to be active.

martinpitt commented on pkg/systemd/terminal.jsx:
```diff
@@ -94,12 +105,12 @@ const _ = cockpit.gettext;
+                            <StatelessSelect id="theme-select" iconClass="fa fa-cog">
+                                <span className="terminal-group">
+                                    <button className={ 'theme-btn dark-theme ' + this.state.darkTheme } onClick={() => this.onThemeChanged('dark-theme') } />
+                                    <button className={ 'theme-btn light-theme ' + this.state.lightTheme } onClick={() => this.onThemeChanged('light-theme') } />
```
While that looks kind of cool, I think we should have a standard vertical selection with theme names. Not everyone might be able to see the fine nuances in the colors in the small selector.

marusak replied:
I will wait for review from @garrett . In the case he agrees with having simple dropdown selection, then I will drop last commits (and add two themes). That also will resolve almost all of your finds, so until then I will not resolve those.

martinpitt commented:
You need to declare the `iconClass` propType, see 
[failing fedora 29 test](https://fedorapeople.org/groups/cockpit/logs/pull-11182-20190218-164848-67e118ee-verify-fedora-29/log.html#183).

marusak commented:
Thank you @garrett for the review. I've dropped last commits and used standard dropdown. See attached screenshots. It would be great if you would prepare some design how we can also add setting of font size and as follow up I can address this as well. Please can you review again? Thank you
![white](https://user-images.githubusercontent.com/12330670/53017475-50415e00-3450-11e9-8021-f4643e0a41f8.png)
![light](https://user-images.githubusercontent.com/12330670/53017476-50415e00-3450-11e9-962e-5d48ef3877de.png)
![dark](https://user-images.githubusercontent.com/12330670/53017477-50d9f480-3450-11e9-834d-8f3b53e0c4a7.png)
![black](https://user-images.githubusercontent.com/12330670/53017478-50d9f480-3450-11e9-81aa-3f1ee88675e6.png)

martinpitt commented on pkg/systemd/terminal.jsx:
```diff
@@ -71,19 +82,31 @@ const _ = cockpit.gettext;
-                             className="btn btn-default pull-right"
-                             onClick={this.onResetClick}>{_("Reset")}</button>
+                        <div>
+                            <span>Appearance:</span>
```
This should be a `label` with `htmlFor="theme-select"`. Also, the string needs to be i18n'ed.

marusak commented:
>  The cursor is invisible in white theme, as it's white-on-white (you can see it if you select over it)

Good catch, this is bug. Will fix it.

> with an empty $LS_COLORS, the colors in black and white are fine, but with light and dark you only get two shades of gray.

This is not really an issue. Because that's how solarized sets colors. Surely we can set different colors, but then it is not gonna be solarized.

marusak commented:
Updated @martinpitt . I also run tests on this for the first (intentional) time.

martinpitt commented:
> Because that's how solarized sets colors

Ah, that's fine then if it's by design. I was a bit concerned about low contrast, but we do have the high-contrast themes now. Thanks!

garrett commented:
Overall, looks good! I did notice a few things:

1. Dropdown isn't accessible. It's impossible to use the keyboard to select items. If it's our custom dropdown, that'll change with #11155.
2. Dropdown changes size depending on length of string. At least for English, adding `min-width: 8rem;` to `#theme-select` would fix this. (7rem would work for English, but considering German, which is usually the one of the longer languages, 8rem is better for the longest translated string... “Schwarz”.)

Item 1 is more important overall (due to being functionality), but not important for this PR to land (as it's a problem elsewhere, and may be fixed elsewhere in the near-future).

Item 2 is less important overall, but an easy 1-line CSS fix that I would like to see land in this PR, if possible.

garrett commented on pkg/systemd/terminal.css:
```diff
@@ -13,9 +13,35 @@ html, body {
+#theme-select {
+    margin-left: 5px;
+    margin-right: 5px;
+    width: auto;
```
```suggestion
    width: auto;
    min-width: 8rem;
```

Add a min-width of 8rem to make the theme selector not shift width when choosing another theme.

marusak commented:
Just updated screenshot how now menu looks like
![darktheme](https://user-images.githubusercontent.com/12330670/53081922-19288680-34fc-11e9-9537-5b652b37b0dc.png)

garrett commented:
Nice. Thanks. I think it's a little wide, but not too wide.

(FWIW: Native dropdowns usually take this kind of thing into consideration.)
